src/output-json-modbus.c

```diff
@@ -0,0 +1,388 @@
+            json_object_set_new(data_dict, key, json_integer(val));
+        }
+        return data_dict;
+    } else {
+        json_t *data_array = json_array();
+        if (data_array == NULL)
+            goto end;
+        for (i = 0; i < quantity; i++) {
+            /* bit by bit reading data pointed by modbustx->data  */
+            val = *(modbustx->data + (i >> 3)) & (1 << (i % 8)) ? 1 : 0;
+            json_array_append_new(data_array, json_integer(val));
+        }
+        return data_array;
+    }
+end:
+    return NULL;
+}
+
+static void JsonModbusLog(json_t *modbusjs, ModbusTransaction *modbustx,
+                          uint8_t use_dict)
+{
+    const char *str_value = ModbusGetFunctionName(modbustx->function);
+
+    json_object_set_new(modbusjs, "transaction_id", json_integer(modbustx->transactionId));
+    json_object_set_new(modbusjs, "length", json_integer(modbustx->length));
+    json_object_set_new(modbusjs, "function", json_string(str_value));
+
+    if (modbustx->function == MODBUS_FUNC_DIAGNOSTIC) {
+        str_value = ModbusGetSubFunctionName(modbustx->subFunction);
+        json_object_set_new(modbusjs, "subfunction", json_string(str_value));
+    }
+
+    if (modbustx->exception) {
+        str_value = ModbusGetExceptionName(modbustx->exception);
+        json_object_set_new(modbusjs, "exception", json_string(str_value));
+    }
+
+    if (modbustx->type & MODBUS_TYP_READ) {
+        json_t *type_read = json_object();
+        if (type_read) {
+            json_object_set_new(type_read, "address", json_integer(modbustx->read.address));
+            json_object_set_new(type_read, "quantity", json_integer(modbustx->read.quantity));
+            if (modbustx->read.quantity && modbustx->data) {
+                json_t *value = NULL;
+                if (modbustx->type  & MODBUS_TYP_BIT_ACCESS_MASK) {
+                    value = JsonModbusLogBits(modbustx, modbustx->read.quantity,
+                                              use_dict);
+                } else {
+                    value = JsonModbusLogRegisters(modbustx, modbustx->read.quantity,
+                                                   use_dict);
+                }
+                if (value) {
+                    json_object_set_new(type_read, "value", value);
+                }
+            }
+            json_object_set_new(modbusjs, "read", type_read);
+        }
+    }
+    if (modbustx->type & MODBUS_TYP_WRITE) {
+        json_t *type_write = json_object();
+        if (type_write) {
+            json_object_set_new(type_write, "address", json_integer(modbustx->write.address));
+            json_object_set_new(type_write, "quantity", json_integer(modbustx->write.quantity));
+            json_object_set_new(type_write, "count", json_integer(modbustx->write.count));
+            if (modbustx->function == MODBUS_FUNC_WRITESINGLEREG) {
+                json_object_set(type_write, "value", json_integer(*(modbustx->data)));
```

Comment:
I think this might be null if the response is an exception. https://github.com/glongo/suricata/blob/modbus-logging-v5/src/app-layer-modbus.c#L532 isn't populating data on exception.

When null, this will segfault.